airflow/models/param.py

```diff
@@ -0,0 +1,208 @@
+        """
+        Override for dictionary's ``getitem`` method. After fetching the key, it would call the
+        resolve method as well on the Param object.
+
+        :param key: The key to fetch
+        :type key: str
+        """
+        param = dict.__getitem__(self, key)
+        return param.resolve(suppress_exception=self.suppress_exception)
+
+    def dump(self) -> dict:
+        """Dumps the ParamsDict object as a dictionary, while suppressing exceptions"""
+        return {k: v.resolve(suppress_exception=True) for k, v in self.items()}
+
+    def update(self, other_dict: dict) -> None:
+        """
+        Override for dictionary's update method.
+        :param other_dict: A dict type object which needs to be merged in the ParamsDict object
+        :type other_dict: dict
+        """
+        try:
+            for k, v in other_dict.items():
+                self.__setitem__(k, v)
+        except ValueError as ve:
+            raise ValueError(f'Invalid input for param {k}: {ve}') from None
+
+    def validate(self) -> None:
+        """Validates & returns all the Params object stored in the dictionary"""
+        resolved_dict = {}
+        try:
+            for k, v in dict.items(self):
+                resolved_dict[k] = v.resolve(suppress_exception=self.suppress_exception)
+        except ValueError as ve:
+            raise ValueError(f'Invalid input for param {k}: {ve}') from None
+
+        return resolved_dict
+
+
+class DagParam:
+    """
+    Class that represents a DAG run parameter & binds a simple Param object to a name within a DAG instance,
+    so that it can be resolved during the run time via {{ context }} dictionary. The ideal use case of this
+    class is to implicitly convert args passed to a method which is being decorated by @dag keyword.
+
+    It can be used to parameterize your dags. You can overwrite its value by setting it on conf
+    when you trigger your DagRun.
+
+    This can also be used in templates by accessing {{context.params}} dictionary.
+
+    **Example**:
+
+        with DAG(...) as dag:
+          EmailOperator(subject=dag.param('subject', 'Hi from Airflow!'))
+
+    :param current_dag: Dag being used for parameter.
+    :type current_dag: airflow.models.DAG
+    :param name: key value which is used to set the parameter
+    :type name: str
+    :param default: Default value used if no parameter was set.
+    :type default: Any
+    """
+
+    def __init__(self, current_dag, name: str, default: Optional[Any] = None):
+        if default:
+            current_dag.params[name] = default
+        self._name = name
+        self._default = default
+
+    def resolve(self, context: Dict) -> Any:
+        """Pull DagParam value from DagRun context. This method is run during ``op.execute()``."""
+        default = self._default
+        if not self._default:
+            default = context['params'][self._name] if self._name in context['params'] else None
```

Comment:
```suggestion
            default = context['params'].get('self._name')
```

Simplified representation